lib/clientv2/watch_e2e_test.go

```diff
@@ -0,0 +1,182 @@
+	numEvents := 10000
+	numWatchers := 1000
+
+	Describe("Test prefix deletion of the datastore", func() {
+		It("should receive watch events for each of the deleted keys", func() {
+			c, err := New(config)
+			Expect(err).NotTo(HaveOccurred())
+
+			be, err := backend.NewClient(config)
+			Expect(err).NotTo(HaveOccurred())
+			be.Clean()
+
+			By("Creating 10000 resources")
+			deleteEvents := []watch.Event{}
+			for ii := 1; ii <= numEvents; ii++ {
+				name := fmt.Sprintf("peer-%08d", ii)
+				peer := &apiv2.BGPPeer{
+					ObjectMeta: metav1.ObjectMeta{Name: name},
+					Spec: apiv2.BGPPeerSpec{
+						PeerIP:   "1.2.3.4",
+						ASNumber: numorstring.ASNumber(ii),
+					},
+				}
+				_, outError := c.BGPPeers().Create(ctx, peer, options.SetOptions{})
+				Expect(outError).NotTo(HaveOccurred())
+				deleteEvents = append(deleteEvents, watch.Event{
+					Type:     watch.Deleted,
+					Previous: peer,
+				})
+			}
+
+			By("Listing all the resources")
+			outList, outError := c.BGPPeers().List(ctx, options.ListOptions{})
+			Expect(outError).NotTo(HaveOccurred())
+			Expect(outList.Items).To(HaveLen(numEvents))
+
+			By("Creating a watcher, watching the current resource version")
+			w, _ := c.BGPPeers().Watch(ctx, options.ListOptions{ResourceVersion: outList.ResourceVersion})
+			testWatcher := testutils.TestResourceWatch(w)
+			defer testWatcher.Stop()
+
+			By("Cleaning the datastore and expecting deletion events for each configured resource (tests prefix deletes results in individual events for each key)")
+			be.Clean()
+			testWatcher.ExpectEvents(apiv2.KindBGPPeer, deleteEvents)
+
+			By("Stopping the watcher and checking for full termination")
+			testWatcher.Stop()
+			Eventually(w.(*watcher).hasTerminated).Should(BeTrue())
+		})
+	})
+
+	Describe("Test constant stream of events whilst closing watcher", func() {
+		It("should handle gracefully closing watchers while events are occurring", func() {
+			c, err := New(config)
+			Expect(err).NotTo(HaveOccurred())
+
+			be, err := backend.NewClient(config)
+			Expect(err).NotTo(HaveOccurred())
+			be.Clean()
+
+			// We create a number of threads in this test.  Use a context as a simple mechanism
+			// for cancelling those threads.
+			finishctx, finish := context.WithCancel(ctx)
+
+			By("Creating an asynchronous events generator")
+			wg := sync.WaitGroup{}
+
+			// Create a goroutine to generate add/delete events.
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				for {
+					if _, err := c.BGPPeers().Create(ctx, &apiv2.BGPPeer{
+						ObjectMeta: metav1.ObjectMeta{Name: "name1"},
+						Spec: apiv2.BGPPeerSpec{
+							PeerIP:   "1.2.3.4",
+							ASNumber: numorstring.ASNumber(12345),
+						},
+					}, options.SetOptions{}); err != nil {
+						panic(err)
+					}
+					time.Sleep(5 * time.Millisecond)
+					if err := c.BGPPeers().Delete(ctx, "name1", options.DeleteOptions{}); err != nil {
+						panic(err)
+					}
+					time.Sleep(5 * time.Millisecond)
```

Comment:
is 5ms enough to check `finishctx.Err()`

Reply:
Not sure I follow.  The 5ms delay is purely so this doesn't tight-loop doing add/delete events.